### PR TITLE
[RTM] Provide a configuration option to define the predefined image sizes

### PR DESCRIPTION
**Pull request by qzminski**
This will allow to configure the image sizes in `config.yml` file as follows:

```yaml
contao:
    image:
        sizes:
            foobar:
                width: 100
                height: 200
                resizeMode: 'box'
                zoom: 100
                cssClass: 'foobar-image'
                densities: '1x, 2x'
                sizes: '100vw'
                items:
                    -
                        width: 50
                        height: 50
                        resizeMode: 'box'
                        zoom: 100
                        cssClass: 'foobar-image'
                        densities: '0.5x, 2x'
                        sizes: '50vw'
                        media: '(max-width: 900px)'
                    -
                        width: 25
                        height: 25
                        resizeMode: 'box'
                        zoom: 100
                        densities: '0.5x, 2x'
                        sizes: '25vw'
                        media: '(max-width: 450px)'
            foobaz:
                width: 300
                height: 400
                # …
```

The labels would be defined in a [Symfony translation file](https://symfony.com/doc/current/translation.html#translation-resource-file-names-and-locations):

```yml
# /translations/image_sizes.en.yml
foobar: Foobar image size
```

Idea came up together with @Toflar.

### CODE REVIEW

File: core-bundle/src/Image/ImageSizes.php [hunk 1]
```diff
@@ -118,6 +131,15 @@ private function loadOptions(): void
                 $imageSize['height']
             );
         }
+
+        foreach ($this->predefinedSizes as $name => $imageSize) {
+            $this->options['image_sizes'][$name] = sprintf(
+                '%s (%sx%s)',
+                $GLOBALS['TL_LANG']['IMAGE_SIZES'][$name] ?? $name,
```
Comment (Toflar) [inline core-bundle/src/Image/ImageSizes.php]: We should use the translator service here, imho :)

**Review by Toflar — CHANGES_REQUESTED** [review]
Awesome! This will allow us to configure image sizes that are versioned and can be deployed to multiple setups 🕺

**Comment by fritzmg** [comment]
How do you reference these in the code (e.g. with `Controller::addImageToTemplate`)?

Just like this?
```
[null, null, 'foobar']
```

**Comment by qzminski** [comment]
@fritzmg yes, exactly like that.

**Comment by fritzmg** [comment]
Nice :). However, what happens if you define something like this?
```yml
contao:
    image:
        sizes:
            box:
                width: 100
                height: 200
                resizeMode: 'box'
 ```

**Comment by qzminski** [comment]
@fritzmg that is a valid point, also relates to the database entries which consists only of the ID. We can either make something prevail or add validation upon container compiling. What do you think guys?

**Comment by Toflar** [comment]
Validate👌🏻

**Comment by qzminski** [comment]
Validation done and you can guys review everything now.

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -153,6 +154,49 @@ function (string $value): string {
...
+                            ->arrayPrototype()
+                                ->children()
+                                    ->integerNode('width')
+                                    ->end()
+                                    ->integerNode('height')
+                                    ->end()
+                                    ->enumNode('resizeMode')
+                                        ->values([ResizeConfigurationInterface::MODE_CROP, ResizeConfigurationInterface::MODE_BOX, ResizeConfigurationInterface::MODE_PROPORTIONAL])
+                                    ->end()
+                                    ->integerNode('zoom')
```
Comment (aschempp) [inline core-bundle/src/DependencyInjection/Configuration.php:215]: should this validate for >=0 <= 100 ?

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php [hunk 3]
```diff
@@ -99,6 +101,32 @@ public function load(array $configs, ContainerBuilder $container): void
...
+            }
+
+            if (in_array($name, [ResizeConfigurationInterface::MODE_BOX, ResizeConfigurationInterface::MODE_PROPORTIONAL, ResizeConfigurationInterface::MODE_CROP], true)) {
+                throw new \RuntimeException(sprintf('Image size name "%s" is reserved and not allowed!', $name));
+            }
+
+            $imageSizes['_' . $name] = $value;
+        }
+
+        $container->setParameter('contao.image.sizes', $imageSizes);
```
Comment (aschempp) [inline core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: There should not be a need to store this as a parameter, we should rather inject it into the correct service directly. Something like 

```php
$def = $container->getDefinition('contao.image_factory');
$def->addMethodCall('setPredefinedSizes', $imageSizes);
```

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -153,6 +154,49 @@ function (string $value): string {
                         ->booleanNode('reject_large_uploads')
                             ->defaultValue(false)
                         ->end()
+                        ->arrayNode('sizes')
+                            ->useAttributeAsKey('name')
```
Comment (aschempp) [inline core-bundle/src/DependencyInjection/Configuration.php:158]: Don't we need a configuration for the `name` attribute? And it must be validated to not be integer, otherwise it would conflict with the database.

File: core-bundle/src/Image/ImageFactory.php [hunk 5]
```diff
@@ -195,21 +208,41 @@ private function createConfig($size, ImageInterface $image): array
...
+                        ->setZoomLevel((int) $imageSize->zoom)
+                    ;
+                }
+
+                return [$config, null];
+            }
+
+            // Predefined sizes
+            if (isset($this->predefinedSizes[$size[2]])) {
+                $imageSize = $this->predefinedSizes[$size[2]];
```
Comment (aschempp) [inline core-bundle/src/Image/ImageFactory.php]: Smells like duplicate code to me. How about

```php

if (is_numeric($size[2])) {
    /** @var ImageSizeModel $imageModel */
    $imageModel = $this->framework->getAdapter(ImageSizeModel::class);
    $imageSize = $imageModel->findByPk($size[2]);
} elseif (isset($this->predefinedSizes[$size[2]])) {
    $imageSize = new ImageSizeModel($this->predefinedSizes[$size[2]]);
}

return [$this->createResizeConfig($imageSize), null];


private function createResizeConfig(ImageSizeModel $imageSize = null)
{
    $config = new ResizeConfiguration();

    if (null !== $imageSize) {
        $config
            ->setWidth((int) $imageSize->width)
            ->setHeight((int) $imageSize->height)
            ->setMode($imageSize->resizeMode)
            ->setZoomLevel((int) $imageSize->zoom)
        ;
    }

    return $config;
}
```

File: core-bundle/src/Resources/config/services.yml [hunk 6]
```diff
@@ -197,6 +197,8 @@ services:
             - '%contao.image.imagine_options%'
             - '%contao.image.valid_extensions%'
             - '%kernel.project_dir%/%contao.upload_path%'
+        calls:
+            - { method: 'setPredefinedSizes', arguments: ['%contao.image.sizes%']}
```
Comment (aschempp) [inline core-bundle/src/Resources/config/services.yml]: This is not necessary when the extension injects the argument

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/src/DependencyInjection/Configuration.php (hunk 4 above)
Comment (qzminski) [reply core-bundle/src/DependencyInjection/Configuration.php:158]: It is validated in `ContaoCoreExtension` class.

Comment (ausi) [reply core-bundle/src/DependencyInjection/Configuration.php:158]: I think you could also use something like this here to validate the name:

```php
->scalarNode('name')
	->validate()
	    ->ifTrue(function($name) {
	        return preg_match('/^\d+$/', (string) $name);
	    })
	    ->thenInvalid('Image size name "%s" cannot contain only digits!')
	    ->ifTrue(function($name) {
	        return in_array($name, [ResizeConfigurationInterface::MODE_BOX, ResizeConfigurationInterface::MODE_PROPORTIONAL, ResizeConfigurationInterface::MODE_CROP], true);
	    })
	    ->thenInvalid('Image size name "%s" is reserved and not allowed!')
	->end()
->end()
```

Not sure what the best practice is here though.

**Comment by qzminski** [comment]
I rebased my changes on #539 :+1:

**Comment by qzminski** [comment]
Here is a screenshot how it looks in the backend:

Zrzut ekranu 2019-07-1 o 11 50 18

**Review by Toflar — APPROVED** [review]
Super nice!

File: core-bundle/src/Image/ImageSizes.php [hunk 7]
```diff
@@ -119,6 +139,19 @@ private function loadOptions(): void
 
         $options = [];
 
+        foreach ($this->predefinedSizes as $name => $imageSize) {
+            if (!isset($options['image_sizes'])) {
+                $options['image_sizes'] = [];
```
Comment (leofeyer) [inline core-bundle/src/Image/ImageSizes.php]: I think this is no longer necessary as I had reverted the changes to `$GLOBALS['TL_CROP']`.

File: core-bundle/src/Resources/contao/templates/backend/be_main.html5 [hunk 8]
```diff
@@ -126,4 +126,4 @@
   <script>Backend.initScrollOffset()</script>
 
 </body>
-</html>
\ No newline at end of file
+</html>
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/templates/backend/be_main.html5]: This change seems unrelated.

File: core-bundle/src/Image/ImageSizes.php [hunk 9]
```diff
@@ -119,6 +139,19 @@ private function loadOptions(): void
...
+            }
+
+            $options['image_sizes'][$name] = sprintf(
+                '%s (%sx%s)',
+                $this->translator->trans('IMAGE_SIZES.' . substr($name, 1), [], 'contao_default') ?: substr($name, 1),
```
Comment (leofeyer) [inline core-bundle/src/Image/ImageSizes.php]: What exactly does this do? There is no `$GLOBALS['TL_LANG']['IMAGE_SIZES']` array.

Comment (qzminski) [reply core-bundle/src/Image/ImageSizes.php]: You are right there is none at the moment because we don't have any predefined sizes. However I would like to have this optional array as a place for labels for those predefined image sizes.

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php [hunk 10]
```diff
@@ -99,6 +101,28 @@ public function load(array $configs, ContainerBuilder $container): void
...
+        $imageSizes = [];
+
+        foreach ($config['image']['sizes'] as $name => $value) {
+            $imageSizes['_' . $name] = $value;
+        }
+
+        $services = ['contao.image.image_sizes', 'contao.image.image_factory', 'contao.image.picture_factory'];
+
+        foreach ($services as $service) {
+            $container->getDefinition($service)->addMethodCall('setPredefinedSizes', [$imageSizes]);
```
Comment (ausi) [inline core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: This is actually a BC break I think because `image_factory` and `picture_factory` have an interface that doesn’t include the `setPredefinedSizes()` method. We would have to check if the services are an instanceof `ImageFactory` and `PictureFactory` and only add the method call then.

Comment (qzminski) [reply core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: Good catch, I have adjusted the code accordingly.

Comment (Toflar) [reply core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: I don't think this is the right way to go. That means I cannot replace it while still have predefined sizes :) Imho we should use `method_exists` just like Symfony does in many places too :)

File: core-bundle/src/Image/ImageSizes.php (hunk 9 above)
Comment (leofeyer) [reply core-bundle/src/Image/ImageSizes.php]: Not sure about this. I would prefer to define the label (or translation key) in the config file.

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php (hunk 10 above)
Comment (qzminski) [reply core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: Adjusted again 😉

File: core-bundle/src/Image/ImageSizes.php (hunk 9 above)
Comment (qzminski) [reply core-bundle/src/Image/ImageSizes.php]: But then you can't define multilingual labels, or can you?

Comment (leofeyer) [reply core-bundle/src/Image/ImageSizes.php]: If you provide a translation key, the Symfony translator will find the corresponding label. This implies to create a language file following the Symfony standard of course.

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 11]
```diff
@@ -153,6 +154,68 @@ function (string $value): string {
...
+                            ->validate()
+                                ->always(function ($value) {
+                                    foreach ($value as $name => $config) {
+                                        if (preg_match('/^\d+$/', (string) $name)) {
+                                            throw new \InvalidArgumentException(sprintf('Image size name "%s" cannot contain only digits!', $name));
+                                        }
+
+                                        if (in_array($name, [ResizeConfigurationInterface::MODE_BOX, ResizeConfigurationInterface::MODE_PROPORTIONAL, ResizeConfigurationInterface::MODE_CROP], true)) {
+                                            throw new \InvalidArgumentException(sprintf('Image size name "%s" is reserved and not allowed (reserved words: %s)!', $name, implode(', ', [ResizeConfigurationInterface::MODE_BOX, ResizeConfigurationInterface::MODE_PROPORTIONAL, ResizeConfigurationInterface::MODE_CROP])));
```
Comment (ausi) [inline core-bundle/src/DependencyInjection/Configuration.php]: Should we also disallow the legacy resize modes here?

left_top, center_top, right_top, left_center, center_center, right_center, left_bottom, center_bottom, right_bottom

File: core-bundle/src/Image/ImageFactory.php [hunk 12]
```diff
@@ -232,6 +249,22 @@ private function createConfig($size, ImageInterface $image): array
         return [$config, $this->getImportantPartFromLegacyMode($image, $size[2])];
     }
 
+    /**
+     * Enhances the resize config by image size settings.
+     */
+    private function enhanceResizeConfig(ResizeConfigurationInterface $config, array $imageSize): void
+    {
+        $config
+            ->setWidth((int) $imageSize['width'])
```
Comment (ausi) [inline core-bundle/src/Image/ImageFactory.php]: We should check if `$imageSize['width']` is set here I think.
```suggestion
            ->setWidth((int) ($imageSize['width'] ?? 0))
```

File: core-bundle/src/Image/ImageFactory.php [hunk 13]
```diff
@@ -232,6 +249,22 @@ private function createConfig($size, ImageInterface $image): array
...
+            ->setHeight((int) $imageSize['height'])
```
Comment (ausi) [inline core-bundle/src/Image/ImageFactory.php]:
```suggestion
            ->setHeight((int) ($imageSize['height'] ?? 0))
```

File: core-bundle/src/Image/ImageFactory.php [hunk 14]
```diff
@@ -232,6 +249,22 @@ private function createConfig($size, ImageInterface $image): array
...
+            ->setZoomLevel((int) $imageSize['zoom'])
```
Comment (ausi) [inline core-bundle/src/Image/ImageFactory.php]:
```suggestion
            ->setZoomLevel((int) ($imageSize['zoom'] ?? 0))
```

File: core-bundle/src/Image/ImageSizes.php [hunk 15]
```diff
@@ -119,6 +139,15 @@ private function loadOptions(): void
 
         $options = [];
 
+        foreach ($this->predefinedSizes as $name => $imageSize) {
+            $options['image_sizes'][$name] = sprintf(
+                '%s (%sx%s)',
+                $this->translator->trans(substr($name, 1), [], 'image_sizes') ?: substr($name, 1),
+                $imageSize['width'],
```
Comment (ausi) [inline core-bundle/src/Image/ImageSizes.php]:
```suggestion
                $imageSize['width'] ?? '',
```

File: core-bundle/src/Image/ImageSizes.php [hunk 16]
```diff
@@ -119,6 +139,15 @@ private function loadOptions(): void
...
+                $imageSize['height']
```
Comment (ausi) [inline core-bundle/src/Image/ImageSizes.php]:
```suggestion
                $imageSize['height'] ?? ''
```

File: core-bundle/src/Image/PictureFactory.php [hunk 17]
```diff
@@ -126,86 +139,115 @@ private function createConfig($size): array
...
 
         if (null !== $imageSize) {
             $resizeConfig
-                ->setWidth((int) $imageSize->width)
-                ->setHeight((int) $imageSize->height)
-                ->setMode($imageSize->resizeMode)
-                ->setZoomLevel((int) $imageSize->zoom)
+                ->setWidth((int) $imageSize['width'])
+                ->setHeight((int) $imageSize['height'])
+                ->setZoomLevel((int) $imageSize['zoom'])
```
Comment (ausi) [inline core-bundle/src/Image/PictureFactory.php]: Same here

File: core-bundle/src/Image/PictureFactory.php [hunk 18]
```diff
@@ -126,86 +139,115 @@ private function createConfig($size): array
...
+            if (isset($imageSize['resizeMode'])) {
+                $resizeConfig->setMode($imageSize['resizeMode']);
+            }
+
             $configItem
                 ->setResizeConfig($resizeConfig)
-                ->setSizes($imageSize->sizes)
-                ->setDensities($imageSize->densities)
+                ->setSizes((string) $imageSize['sizes'])
+                ->setDensities((string) $imageSize['densities'])
```
Comment (ausi) [inline core-bundle/src/Image/PictureFactory.php]: Same here

**Review by ausi — CHANGES_REQUESTED** [review]

File: core-bundle/src/DependencyInjection/Configuration.php (hunk 11 above)
Comment (leofeyer) [reply core-bundle/src/DependencyInjection/Configuration.php]: These modes are part of `ResizeConfigurationInterface::MODE_CROP`, aren't they? We do not have to disallow them here IMHO.

Comment (ausi) [reply core-bundle/src/DependencyInjection/Configuration.php]: But what if you name your image size `left_top` and then use `[0, 0, 'left_top']`, which size should then be used?

File: core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php [hunk 19]
```diff
@@ -1821,4 +1821,75 @@ public function testRegistersTheImageTargetPath(): void
...
+            $this->container
+        );
+
+        foreach ($services as $service) {
+            $this->assertTrue($this->container->getDefinition($service)->hasMethodCall('setPredefinedSizes'));
+        }
+    }
+
+    /**
+     * @expectedException \RuntimeException
```
Comment (leofeyer) [inline core-bundle/tests/DependencyInjection/ContaoCoreExtensionTest.php]: Please use `$this->expectException()` instead.

File: core-bundle/tests/Image/ImageSizesTest.php [hunk 20]
```diff
@@ -186,4 +197,15 @@ private function expectExampleImageSizes(): void
             ],
         ]);
     }
+
+    /**
+     * Adds expected example predefined image sizes.
```
Comment (leofeyer) [inline core-bundle/tests/Image/ImageSizesTest.php]: This comment is superfluous.

File: core-bundle/src/DependencyInjection/ContaoCoreExtension.php [hunk 21]
```diff
@@ -99,6 +100,30 @@ public function load(array $configs, ContainerBuilder $container): void
         ;
     }
 
+    /**
+     * Validates and sets the "contao.image.sizes" parameter.
+     */
+    private function setPredefinedImageSizes(array $config, ContainerBuilder $container): void
+    {
+        if (!isset($config['image']['sizes']) || count($config['image']['sizes']) === 0) {
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/ContaoCoreExtension.php]: CS: `0 === \count($config['image']['sizes'])`

You should probably run the CS fixer. 😄

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/DependencyInjection/Configuration.php (hunk 11 above)
Comment (leofeyer) [reply core-bundle/src/DependencyInjection/Configuration.php]: You are right. 🦊

**Review by ausi — APPROVED** [review]

File: core-bundle/src/DependencyInjection/Configuration.php [hunk 22]
```diff
@@ -153,6 +154,104 @@ function (string $value): string {
...
+                        ->arrayNode('sizes')
+                            ->useAttributeAsKey('name')
+                            ->validate()
+                                ->always(
+                                    static function (array $value): array {
+                                        static $reservedImageSizeNames = [
+                                            ResizeConfigurationInterface::MODE_BOX,
+                                            ResizeConfigurationInterface::MODE_PROPORTIONAL,
+                                            ResizeConfigurationInterface::MODE_CROP,
+                                            'left_top',
```
Comment (leofeyer) [inline core-bundle/src/DependencyInjection/Configuration.php:166]: @ausi I wonder if we should add a `ResizeConfigurationInterface::CROP_MODES` constant so we do not have to duplicate this code?

File: core-bundle/src/Image/ImageSizes.php (hunk 15 above)
Comment (leofeyer) [reply core-bundle/src/Image/ImageSizes.php]: @qzminski Why did you decide against @ausi's suggestion?

Comment (qzminski) [reply core-bundle/src/Image/ImageSizes.php]: @leofeyer I added it in wrong place 🙈see 8555e9796

Comment (leofeyer) [reply core-bundle/src/Image/ImageSizes.php]: There are some more places where @ausi requested to change this. I have unresolved all matching conversations so you can track them more easily.

Comment (leofeyer) [reply core-bundle/src/Image/ImageSizes.php]: Ah, no, my bad. You have already wrapped those places with `isset()`. Sorry.

**Review by leofeyer — APPROVED** [review]

File: core-bundle/src/Image/ImageSizes.php (hunk 15 above)
Comment (qzminski) [reply core-bundle/src/Image/ImageSizes.php]: I decided to use `isset($val) ? (int) $val … : 0` instead of `(int) ($val ?? 0)` because – to be precise – there is no need to cast `0` to `int`.

**Comment by leofeyer** [comment]
Thank you @qzminski.
